### ISSUE TITLE
Kindle Scribe: Random finger touch unresponsiveness

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#11482]
* KOReader version: v2024.01-63
* Device: Kindle Scribe (5.16.2)

#### Issue
I got some weird freezes in some case, but inconsistently.
1) First one was when I had a vector-image heavy PDF file (scientific article) which I asked about recently. I don't remember what gesture I used to trigger the freeze.
2) Two times when I did my diagonal swipe with two fingers in the direction of top right corner to summon bookmarks. I did a hard restart after each. This was surprisingly while I was reading an FB2 book!

Symptoms are the same. I do a gesture and KOReader freezes to touch.
I can put it to sleep, screensaver shows up alright.
I press power button again, and instead of saying "Use your gesture to unlock", it says "Unlock gesture is not setup, tap to unlock" (don't remember exact wording). Then it is frozen again.
I didn't have access to my laptop wtih SSH in all three times, but now I can't reproduce it neither.

Is there any pointers in the log? In the second freeze in (2) I enabled verbose logs.
I kept them on now, to be able to catch it again.

Thanks!

##### `crash.log` (if applicable)
[crash_FREEZE.log.txt](https://github.com/koreader/koreader/files/14335266/crash_FREEZE.log.txt)
[crash_FREEZE_verbose.log.txt](https://github.com/koreader/koreader/files/14335267/crash_FREEZE_verbose.log.txt)

**Comment by mergen3107** [comment]
I have a suspicion that this is touch driver is acting up. I remembered another time when it froze like this last week: when I hit a link in the above mentioned PDF and wanted to go back via my two finger swipe left gesture. 

Maybe Kindle's multitouch implementation is a hit or miss?

**Comment by mergen3107** [comment]
I suspect two-finger swipes are somehow tripping the touch input device. 

I see some messages in first log like “Restoring Kindle input” and “Inhibiting touch input” or something like that. 

Then I also don’t know if those “closing touch device fd” are relevant or not, but maybe not - there appear before a usual clean exit with code 0. 

@NiLuJe can you please suggest where should I dig? I think it might be related with:
1) Two-finger swipes / multiswipe detection racing to determine the gesture
2) Two-finger swipe / short diagonal swipe or other gestures racing with each other with no definitive resolution
3) Two-finger swipe just being wonky by themselves. 

I have yet to reproduce it live with SSH on. (As if with verbose logs this issue is not reproducible, but I must be tripping).

Sorry, I noticed that verbose log file is empty. I’ll try to reupload later if it’s still there on Kindle.

**Comment by Frenzie** [comment]
You mean that this "want to restart" widget wasn't drawn?

```
 Usable data from seed files: 87.238384%
 Renaming temp file
 Fetching remaining blocks
 Downloading from http://ota.koreader.rocks/koreader-kindle-v2024.01-80-gec98c6334_2024-02-28.targz
 optimized ranges, old requests count 2568, new requests count 18
 
 
#################--- 87.2% 
#################--- 87.3% 8.2 kBps 15:01 ETA   
#################--- 87.6% 319.5 kBps 1:11 ETA   
#################--- 88.3% 532.6 kBps 0:32 ETA   
#################--- 88.7% 707.1 kBps 0:31 ETA   
#################--- 88.7% 812.5 kBps 0:38 ETA   
#################--- 89.1% 922.8 kBps 0:35 ETA   
#################--- 89.2% 910.4 kBps 0:39 ETA   
#################--- 89.5% 906.8 kBps 0:37 ETA   
##################-- 90.8% 919.8 kBps 0:23 ETA   
##################-- 92.3% 1054.5 kBps 0:13 ETA   
##################-- 93.6% 1096.4 kBps 0:09 ETA   
##################-- 94.9% 1117.3 kBps 0:06 ETA   
###################- 96.4% 1118.0 kBps 0:04 ETA   
###################- 97.8% 1105.6 kBps 0:02 ETA   
###################- 98.1% 1117.3 kBps 0:01 ETA   
###################- 98.3% 1112.6 kBps 0:01 ETA   
###################- 100.0% 1192.4 kBps 0:00 ETA   
#################### 100.0% 1027.8 kBps DONE    

 Verifying downloaded file
 checksum matches OK
 used 88438784 local, fetched 18271024
 02/29/24-21:02:50 DEBUG ImageWidget: _render'ing resources/icons/mdlight/notice-question.svg 77 77 
02/29/24-21:02:50 DEBUG ImageWidget: initial offsets 0 0 
02/29/24-21:02:50 DEBUG show widget: table: 0xb5127f70 
02/29/24-21:02:50 DEBUG setDirty nil from widget table: 0xb5127f70 w/ NO region; dithering: nil 
02/29/24-21:02:50 DEBUG setDirty via a func from widget table: 0xb5127f70 
02/29/24-21:02:50 DEBUG not painting 1 covered widget(s) 
02/29/24-21:02:50 DEBUG painting widget: table: 0xb5127f70 
02/29/24-21:02:50 DEBUG blitFrom 261 1157 0 0 77 77 
02/29/24-21:02:50 DEBUG _refresh: Enqueued ui update for region 238 1134 1383 212 dithering: false 
02/29/24-21:02:50 DEBUG triggering refresh {
  dither = false,
  mode = "ui",
  region = "1383x212+238+1134"
} --[[table: 0xb1d0daf8]] 
```

**Comment by NiLuJe** [comment]
> So, it turns out verbose log was 0 KB on Kindle as well.

Whut?

That's not possible, even a non-verbose log will have *some* content in it.

That would imply severe filesystem shenanigans to me (or fuse proxy, if it's a pre-MTP FW).

**Comment by NiLuJe** [comment]
That last log looks perfectly fine, on the other hand, so that's sort of strange. I'd have been curious to know what it was stuck on (e.g., strace it; hitting `S` with the process highligted in `htop` will do that).

**Comment by mergen3107** [comment]
> So, it turns out verbose log was 0 KB on Kindle as well.

@NiLuJe 
First, this file I lost was a result of me renaming crash.log to that new name in kterm (because I only had Kindle on hand at that time on a train). 

First renaming was OK (non-verbose log), second was not (verbose log).

I'll try to htop it, thanks!

**Comment by mergen3107** [comment]
So, to make it clear: crash.log itself was populated alright. Maybe then kterm messed up?

**Comment by mergen3107** [comment]
I did it via `cp file1 file2`

**Comment by mergen3107** [comment]
@NiLuJe 
For a start, here is a strace from `htop -p $(pidof reader.lua)` (then pressed `s`) when I exit normally, FWIW:
![TRACE_OK](https://github.com/koreader/koreader/assets/6279855/74338b44-7006-41cc-a280-1b93901b1a84)
[TRACE_OK.txt](https://github.com/koreader/koreader/files/14483161/TRACE_OK.txt)

Meanwhile I am trying to catch the same freeze with strace open.
However, the strace window stays at the first line like this the whole time:
```
Trace of process 10344 - /mnt/us/koreader/luajit│./luajit ./reader.lua

strace: Process 10344 attached
09:37:12.154993 read(3,
```

Do you expect it to add more details when I get a freeze?
Please guide me here, since I have never used strace before :D

**Comment by mergen3107** [comment]
I haven't been able to catch another freeze yet (moved to 2024.03 stable), but I got an idea.

Can it be that it is not the `reader.lua` that freezes, but only the input module breaks somehow?

I remember @NiLuJe said that the touch control is exclusive, so once KOReader takes the control, Kindle native can't. But what if Kindle UI still tries to gain the control back and succeeds in some cases?

**Comment by mergen3107** [comment]
Is there a way I can send a command (say, "next page") via ssh in KOReader?

**Comment by NiLuJe** [comment]
> But what if Kindle UI still tries to gain the control back and succeeds in some cases?

It can't (at least not from userland). If the input device is a virtual or uinput one and the state of it changes, that's another kettle of fish, though.

Regardless, a strace will tell what's happening in more details.

> Is there a way I can send a command (say, "next page") via ssh in KOReader?

Dunno if something like that made it to @poire-z's HTTP plugin, I haven't had time to catch up quite yet ;).

**Comment by Frenzie** [comment]
You can send events, so the short answer is yes.

**Comment by mergen3107** [comment]
> If the input device is a virtual or uinput

I think we figured out that these `/dev/input/eventX` are not virtual ones, right?
Because virtual ones are in `/sys/class/input/`

**Comment by NiLuJe** [comment]
There may still be a way for something to stop the kernel from reporting events at all (e.g., like what happened with the haptic buttons on the Voyage), dunno ;).

**Comment by mergen3107** [comment]
Sounds like a lot of fun

**Comment by mergen3107** [comment]
I guess now I have a mild PTSD from Kobo Clara HD issues, when it turned out to be… the magnet.

I realized, that those cases of frozen touch were either on the train, airplane, or classroom, where I used Kindle with its magnetic cover semi-folded, so that Kindle stays upright.

I guess, I'll go and try to play around with magnet now :D

(This might explain everything to me expect why Kindle UI was the only one to dismiss finger touches)

**Comment by mergen3107** [comment]
Magnet also doesn't explain why KOReader why showing "No gesture to unlock KOReader, tap to exit" during these first freezes. This looks more like a storage lock up.

**Comment by mergen3107** [comment]
@NiLuJe 
Here is the `dmesg` between I caught the issue (very last time) and some time after (no luck it catching it again. Can you please see if there is anything hinting to the touch/input failures?
[dmesg.txt](https://github.com/koreader/koreader/files/14530356/dmesg.txt)

The only `error` keyword I see about the gyroscope:
```
[    5.872866] .(1)[597:modprobe][E]3<BS_LOG><bma4xy_i2c_read><94>I2C xfer error
[    5.872888] .(1)[597:modprobe][E]3<BS_LOG><bma4xy_probe><2649>Bosch Sensortec Device 1 CANNOT be detected!
[    5.872895] .(1)[597:modprobe][E]3<BS_LOG><bma4xy_i2c_probe><250>BMA4xy Probe failed with err: -2
[    5.872933] .(1)[597:modprobe]bma4xy_acc: probe of 0-0019 failed with error -2
```
and the other is about uart pins:
```
[    0.376105] .(0)[1:swapper/0]mt6577-uart 11003000.serial: Error applying setting, reverse things back
[    0.376126] .(0)[1:swapper/0]mt6577-uart: probe of 11003000.serial failed with error -22
```

**Comment by NiLuJe** [comment]
Nope. Then again, there was no guarantee there'd be anything visible in there to begin with ;).

**Comment by mergen3107** [comment]
So, I still haven't caught that issue again reading everyday too, so frustrated. 

However it seems to be resolved after 2024.3.1-1 (pen slots shenanigan resolved). 

Thank you @NiLuJe if that's what solved it :D 

Closing until I catch it again.

**Comment by mergen3107** [comment]
For crossref, 2024.3.1-1 is this: #11520
